### PR TITLE
fix: checks pod status after creation

### PR DESCRIPTION
**Pull request by d2lam**
Currently, pod creation returns `201`, but afterward, it might fail at initcontainer, or after that. However, the error doesn't get caught because because `_start` already returns. So the status of the build stays as `QUEUED` forever even though the pod failed. 

This PR:
- checks the pod status after it's created. If the status is `failed` or `unknown`, it will throw error. The queue worker will catch this, retry, and report back to API that this build failed (https://github.com/screwdriver-cd/queue-worker/blob/master/index.js#L90-L91). 
- adds random string to pod name so that the queue can recreate the pod (currently unable to recreate pod if name is the same)

Resources:
https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase
https://v1-6.docs.kubernetes.io/docs/api-reference/v1.6/#read-status-69

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -84,6 +82,30 @@ class K8sVMExecutor extends Executor {
...
+                    return this.breaker.runCommand(statusOptions);
+                }))
+            .then((resp) => {
+                if (resp.statusCode !== 200) {
+                    throw new Error(`Failed to get pod status: ${JSON.stringify(resp.body)}`);
+                }
+
+                const status = resp.body.status.phase.toLowerCase();
+
+                if (status === 'failed' || status === 'unknown') {
```
Comment (jithin1987) [inline index.js]: what if it's still queued ? We need to poll till it's status changes. I have a related requirement https://partner.git.corp.yahoo.com/Screwdriver/engine/issues/920

Comment (minz1027) [reply index.js]: Can probably use `requestretry` pkg for making the status call.
https://github.com/screwdriver-cd/screwdriver/blob/master/features/support/world.js#L6

File: index.js [hunk 2]
```diff
@@ -84,6 +82,30 @@ class K8sVMExecutor extends Executor {
                     throw new Error(`Failed to create pod: ${JSON.stringify(resp.body)}`);
                 }
 
+                return resp.body.metadata.name;
+            })
+            .then(podname => new Promise(resolve => setTimeout(resolve, 3000))
```
Comment (petey) [inline index.js]: Neat little delay trick, but should move the magic number to a constant somewhere else.

File: index.js (hunk 1 above)
Comment (FenrirUnbound) [reply index.js]: We've done this kind of "wait and poll" in the functional tests

Comment (FenrirUnbound) [reply index.js]: We still have to implement `blocked_by`, which will affect the "is still queued" logic.

**Comment by FenrirUnbound** [comment]
The PR says that it's a 5 second delay but the code has 3 seconds. Which is the correct value?

File: index.js (hunk 2 above)
Comment (jithin1987) [reply index.js]: use `RETRYDELAY` instead of `3000`

File: index.js [hunk 3]
```diff
@@ -84,6 +98,36 @@ class K8sVMExecutor extends Executor {
...
+                    };
+
+                    return this.breaker.runCommand(statusOptions);
+                }))
+            .then((resp) => {
+                if (resp.statusCode !== 200) {
+                    throw new Error(`Failed to get pod status: ${JSON.stringify(resp.body)}`);
+                }
+
+                console.log(resp);
```
Comment (jithin1987) [inline index.js]: remove log

File: index.js [hunk 4]
```diff
@@ -84,6 +98,36 @@ class K8sVMExecutor extends Executor {
...
+
+                const status = resp.body.status.phase.toLowerCase();
+
+                if (status === 'failed' || status === 'unknown') {
```
Comment (jithin1987) [inline index.js]: I was looking at documentation for `requestretry`. This won't do what we want. We need to poll status even on `200` success requests. AFAIK https://www.npmjs.com/package/requestretry retries on error. What we are polling is on success `200` requests until status changes to `failed/running`

File: index.js (hunk 3 above)
Comment (d2lam) [reply index.js]: Still working on this PR...

File: index.js (hunk 4 above)
Comment (d2lam) [reply index.js]: I think that's the default case. If retryStrategy is defined, it should use the retryStrategy:
https://github.com/FGRibreau/node-request-retry/blob/master/index.js#L85

File: config/pod.yaml.tim [hunk 5]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 kind: Pod
 metadata:
-  name: "{{build_id_with_prefix}}"
+  name: "{{pod_name}}"
```
Comment (jithin1987) [inline config/pod.yaml.tim]: it used to be very easy to describe or see pod logs eg; `kubectl describe pod **12115** --namespace=cd-screwdriver-k8s-build` is that going to change ?

File: test/index.test.js [hunk 6]
```diff
@@ -39,7 +41,8 @@ describe('index', function () {
     });
 
     beforeEach(() => {
-        requestMock = sinon.stub();
+        requestRetryMock = sinon.stub();
+        requestRetryMock = sinon.stub();
```
Comment (jithin1987) [inline test/index.test.js]: duplicate line.

Comment (d2lam) [reply test/index.test.js]: Fixed

File: config/pod.yaml.tim (hunk 5 above)
Comment (d2lam) [reply config/pod.yaml.tim]: We can always do a `k get pods | grep 12115`

Comment (FenrirUnbound) [reply config/pod.yaml.tim]: Could we add another label that's like `build: {{build_id}}`?

Comment (FenrirUnbound) [reply config/pod.yaml.tim]: Nevermind, I see that there's a label for `sdbuild`.

**Comment by FenrirUnbound** [comment]
Just kidding. I guess the description dropped the intended timeout value

**Pull request merged**